In [16]:
import pandas as pd
import statsmodels.api as sm
from scipy import stats

Import bikeInfoMerged csv file from the previous notebook.

In [17]:
df_bike = pd.read_csv('/Users/nikiyaw/Desktop/bikeInfoMerged.csv', sep=';', index_col=0)
df_bike

bikeStationName   Latitude  Longitude  totalBikeSlots  \
0             Curtatone  43.321600  11.327948              16   
1            S Agostino  43.314554  11.331021              10   
2            S Girolamo  43.316539  11.337212               7   
3              Fortezza  43.320523  11.323225              10   
4             Antiporto  43.328983  11.321913              10   
5             Due Ponti  43.317187  11.354256              12   
6                Napoli  43.341757  11.305010               6   
7              Ospedale  43.343243  11.326733              14   
8            Acquacalda  43.336999  11.300975              10   
9            Ravacciano  43.323781  11.338891              10   
10            S Miniato  43.347800  11.326046              10   
11         Terminal Bus  43.332288  11.324176              12   
12            Vico Alto  43.341445  11.312348               8   
13   Casa dell'Ambiente  43.326764  11.333702               6   
14              Frajese  43.309896  11.315263              10   
15              Mercato  43.317315  11.332998              12   
16                 Tufi  43.296856  11.334854               9   
17          Calamandrei  43.343407  11.322849              10   
18            Petriccio  43.334640  11.304803              10   

    F_numberOfCafes  F_averageRatings  Y_numberOfCafes  Y_averageRatings  
0              10.0              7.71             14.0          4.035714  
1              10.0              7.71             13.0          4.000000  
2              10.0               NaN             13.0          4.000000  
3              10.0               NaN             14.0          4.035714  
4              10.0               NaN              5.0          4.100000  
5               6.0               NaN              0.0               NaN  
6               7.0               NaN              1.0          3.000000  
7               0.0               NaN              0.0               NaN  
8               6.0               NaN              1.0          3.000000  
9              10.0               NaN             13.0          4.000000  
10              0.0               NaN              0.0               NaN  
11             10.0               NaN              2.0          3.750000  
12              7.0               NaN              1.0          3.000000  
13             10.0               NaN             11.0          4.045455  
14              5.0               NaN              0.0               NaN  
15             10.0              7.71             13.0          4.000000  
16              0.0               NaN              1.0          4.000000  
17              0.0               NaN              1.0          3.000000  
18              6.0               NaN              2.0          3.000000

Build a regression model.
Provide model output and an interpretation of the results.

We will be creating two simple linear regression models using totalBikeSlots as the dependent variable and F_numberOfCafes & Y_numberOfCafes as the independent variables.

In [18]:
X = df_bike
y = df_bike['totalBikeSlots']

X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X['F_numberOfCafes'])

model_1 = lin_reg.fit()
print_model_1 = model_1.summary()
print(print_model_1)

                                 OLS Regression Results                                
Dep. Variable:         totalBikeSlots   R-squared (uncentered):                   0.696
Model:                            OLS   Adj. R-squared (uncentered):              0.679
Method:                 Least Squares   F-statistic:                              41.14
Date:                Mon, 05 Jun 2023   Prob (F-statistic):                    4.88e-06
Time:                        10:00:50   Log-Likelihood:                         -60.130
No. Observations:                  19   AIC:                                      122.3
Df Residuals:                      18   BIC:                                      123.2
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                      coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------

/Users/nikiyaw/anaconda3/lib/python3.10/site-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


R-squared reflects the fit of the model, which means the model is capable of explaining only 69.6% of the patterns in the data. 
The regression output shows a p-value of 0, which means that the probability of the realtionship between the F_numberOfCafes and totalBikeSlots being solely due to natural variation is 0. The F_numberOfCafes does impact the totalBikeSlots. 
The coefficient value of 1.1229 shows that the increase of F_numberOfCafes will have a positive impact on the totalBikeSlots.

In [19]:
X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X['Y_numberOfCafes'])

model_2 = lin_reg.fit()
print_model_2 = model_2.summary()
print(print_model_2)

                                 OLS Regression Results                                
Dep. Variable:         totalBikeSlots   R-squared (uncentered):                   0.459
Model:                            OLS   Adj. R-squared (uncentered):              0.428
Method:                 Least Squares   F-statistic:                              15.25
Date:                Mon, 05 Jun 2023   Prob (F-statistic):                     0.00104
Time:                        10:00:51   Log-Likelihood:                         -65.602
No. Observations:                  19   AIC:                                      133.2
Df Residuals:                      18   BIC:                                      134.1
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                      coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------

/Users/nikiyaw/anaconda3/lib/python3.10/site-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


R-squared reflects the fit of the model, which means the model is capable of explaining only 45.9% of the patterns in the data. 
The p-value is 0.001, which means that the probability of the realtionship between the F_numberOfCafes and totalBikeSlots being solely due to natural variation is 0.001. The F_numberOfCafes does impact the totalBikeSlots. 
The coefficient value of 0.8753 shows that the increase of F_numberOfCafes will have a positive impact on the totalBikeSlots.

We are obtaining data from two different API sources throughout this exercise, Foursqaure and Yelp. The target variable is the number of cafes within a radius of 1km from all the bike stations. Let's compare if the outcome obtained from both APIs are the same. 

We will use the t-tests to investigate whether the results from both API are the same, by comparing F_numberOfCafes & Y_numberOfCafes. 

In [20]:
df_bike['F_numberOfCafes'].describe()

count    19.000000
mean      6.684211
std       3.958853
min       0.000000
25%       5.500000
50%       7.000000
75%      10.000000
max      10.000000
Name: F_numberOfCafes, dtype: float64

In [21]:
F_mean = df_bike['F_numberOfCafes'].mean()
F_mean

6.684210526315789

In [22]:
df_bike['Y_numberOfCafes'].describe()

count    19.000000
mean      5.526316
std       5.994149
min       0.000000
25%       1.000000
50%       2.000000
75%      13.000000
max      14.000000
Name: Y_numberOfCafes, dtype: float64

In [23]:
Y_mean = df_bike['Y_numberOfCafes'].mean()
Y_mean

5.526315789473684

In [29]:
stat, p = stats.ttest_ind(F_mean, Y_mean)
p

/var/folders/mg/9c8_3q293sgg37fxt_551_tm0000gn/T/ipykernel_3010/912582688.py:1: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  stat, p = stats.ttest_ind(F_mean, Y_mean, equal_var=False)


nan

Equal variance test

In [25]:
F_std = df_bike['F_numberOfCafes'].std()**2
Y_std = df_bike['Y_numberOfCafes'].std()**2
print([F_std, Y_std])

[15.672514619883044, 35.929824561403514]


In [28]:
stat, p = stats.levene(F_mean, Y_mean)
p

TypeError: object of type 'numpy.float64' has no len()

# Stretch

How can you turn the regression model into a classification model?